### ISSUE TITLE
Drawing text with a non-opaque color can sometimes reveal internal glyph overlaps

### ISSUE THREAD

**Issue by flar** [flutter/flutter#123042]
Recently we fixed some bugs in the engine where applying an opacity widget to text might cause internal glyph overlaps to reveal themselves (see https://github.com/flutter/flutter/issues/122266 for example). Rather than applying such opacity as a saveLayer, the engine had been applying it by changing the color used to render the text to a less-opaque version.

While it reduced the number of developer surprises to not equate "alpha saveLayer" with "translucent coloring" in that case, a developer might choose to use a translucent color themselves and they'd see the same overlaps. Such behavior might surprise them as the same result does not occur when they apply a translucent color to a path rendering that has overlaps in it.

We should probably treat any given text operation as an atomic rendering and protect it from overlap similar to how paths are protected from overlap effects during rendering (not similar by way of implementation, but similar by way of expected output).

**Comment by chinmaygarde** [comment]
This seems like working as intended right? The implications of doing this the other way would be severely less performant text rendering.

Not sure how much of a problem this is in the real world. Doing this performantly by default with the option to switch to a save layer based solution seems like the way to go.

**Comment by flar** [comment]
> This seems like working as intended right? The implications of doing this the other way would be severely less performant text rendering.

I don't agree that this is working as intended. It may be working as our implementation team wants to put in the effort, but what does the developer expect from these operations?

Drawing a block of text.
Drawing a complex path, perhaps even a path representing a block of text.

Would the typical developer think of those as operations that have different behaviors? And yet, one of those will fade perfectly as you reduce the alpha of the color and the other will start showing overdraw.

The implications of doing this the other way around should take into account being able to detect glyph overlap in the text rendering code - or to do it in a way that naturally avoids the overdraw. The rendering code has access (or is closer to having access) to the tools to figure this out, but it is difficult if not impossible for the developer to do.

> Not sure how much of a problem this is in the real world. Doing this performantly by default with the option to switch to a save layer based solution seems like the way to go.

One could say the same about paths. How often do they overlap? And yet we protect them from overlap overdraw because that is less surprising to the developer.

**Comment by svetlin-baraka** [comment]
Hi, we face this issue with Arabic, which is unacceptable behavior as our app targets the Gulf states. Native iOS and Android platforms don't have this issue.

<img width="207" alt="image" src="https://github.com/user-attachments/assets/58719dad-7f33-4aa1-a3e4-8d93cbc88d01">
